bundledatasrc.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package charm
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/juju/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type fieldPresenseMap map[interface{}]interface{}
+
+func (fpm fieldPresenseMap) fieldPresent(fieldName string) bool {
+	_, exists := fpm[fieldName]
+	return exists
+}
+
+func (fpm fieldPresenseMap) forField(fieldName string) fieldPresenseMap {
+	v, exists := fpm[fieldName]
+	if !exists {
+		return nil
+	}
+
+	asMap, valid := v.(fieldPresenseMap)
+	if !valid {
+		panic(errors.Errorf("field map entry %q does not point to a nested field presense map", fieldName))
+	}
+	return asMap
+}
+
+// bundleDataPart combines a parsed BundleData instance with a nested map that
+// can be used to discriminate between fields that are missing from the data
+// and those that are present but defined to be empty.
+type bundleDataPart struct {
+	data        *BundleData
+	presenseMap fieldPresenseMap
+}
+
+// BundleDataSource is implemented by types that can parse bundle data into a
+// list of composable parts.
+type BundleDataSource interface {
+	read() ([]*bundleDataPart, error)
+}
+
+// FileBundleDataSource is a BundleDataSource that reads bundle data from a file.
+type FileBundleDataSource string
+
+// read implements BundleDataSource.
+func (s FileBundleDataSource) read() ([]*bundleDataPart, error) {
```

Comment:
Arguably this is a filename bundledatasource because if you had a File that fits as an io.Reader and you would just use StreamBundleDataSource, right?
It seems a bit odd to me that FileBundleDataSource *is a* string while StreamBundleDataSource *has a* reader. I'd rather they were consistent about whether they are actually a struct or not.

Reply:
Also a good observation. As I will be moving the "is-this-a-bundle-or-a-dir-or-an-archive" logic into this type I will also be renaming it to something like a "LocalPathBundleDataSource" and convert it into a struct as we also need to track the base-path for resolving includes.